### PR TITLE
Mesh 1052/Benchmarking base setup

### PR DESCRIPTION
**Pull request by maxsam4**
You may ignore everything in `/bin` directory. That's 80% substrate code and 20% my modifications to adapt it for Polymesh and a few optimizations.

This PR just sets up the base for compiled WASM and benchmarks. The actual benchmarks will come later.

### CODE REVIEW

File: pallets/runtime/common/src/lib.rs [hunk 1]
```diff
@@ -45,10 +48,22 @@ pub type NegativeImbalance<T> =
...
     pub const MaximumBlockWeight: Weight = 2 * WEIGHT_PER_SECOND;
     pub const AvailableBlockRatio: Perbill = Perbill::from_percent(75);
     pub const MaximumBlockLength: u32 = 5 * 1024 * 1024;
+    /// 20 ms is needed to create a block
+    pub const BlockExecutionWeight: Weight = 20 * WEIGHT_PER_MILLIS;
+    // 0.5 ms is needed to process an empty extrinsic
+    pub const ExtrinsicBaseWeight: Weight = 500 * WEIGHT_PER_MICROS;
+    /// When the read/writes are cached, they take 25, 10 microseconds
+    /// but when they are uncached, they take 250, 450 microseconds.
+    /// Most of the read/writes will be ached in production
```
Comment (vkomenda) [inline pallets/runtime/common/src/lib.rs:61]:
```suggestion
    /// Most read/writes are cached in production
```

File: pallets/runtime/common/src/lib.rs [hunk 2]
```diff
@@ -45,10 +48,22 @@ pub type NegativeImbalance<T> =
...
+    /// so we are taking a defensive middle number here
```
Comment (vkomenda) [inline pallets/runtime/common/src/lib.rs:62]:
```suggestion
    /// so we are taking a defensive middle number here.
```

File: Cargo.toml [hunk 3]
```diff
@@ -11,6 +11,8 @@ path = "src/main.rs"
 
 [workspace]
 members = [
+    "bin/bench",
```
Comment (vkomenda) [inline Cargo.toml:14]: Would it be possible to use these bench and testing crates from the Substrate repo?

Comment (maxsam4) [reply Cargo.toml:14]: Nope, the things that can be used from substrate repo are being used directly but these 3 packages required changes to be made. About 20% of these is our custom code.

File: pallets/identity/src/lib.rs [hunk 4]
```diff
@@ -2185,10 +2192,20 @@ impl<T: Trait> IdentityTrait<T::AccountId> for Module<T> {
         });
     }
 
-    // Provides the DID status for the given DID
+    // /Provides the DID status for the given DID
```
Comment (adamdossa) [inline pallets/identity/src/lib.rs:2195]: Looks like a minor typo with a space between the "/"

**Review by adamdossa — APPROVED** [review]

**Review by satyamakgec — APPROVED** [review]

File: bin/executor/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,59 @@
+[package]
+name = "node-executor"
```
Comment (vkomenda) [inline bin/executor/Cargo.toml:2]: Would it be possible to use the Parity crate `node-executor` which is the member of the `substrate` monorepo?

**Review by vkomenda — APPROVED** [review]
It would be great if the copy-pasted code in `bin` could be instead used as external dependencies. If that is not possible, I'm OK with merging this.

File: bin/executor/Cargo.toml (hunk 5 above)
Comment (maxsam4) [reply bin/executor/Cargo.toml:2]: Nope. On L31 of  `bin/executor/src/lib.rs`, I had to inject our WASM host functions to the executor.

**Comment by maxsam4** [comment]
> It would be great if the copy-pasted code in `bin` could be instead used as external dependencies. If that is not possible, I'm OK with merging this.

Unfortunately, that's not possible. That being said, stuff in the `bin` is not required for compilation of the actual Polymesh binary. It is used only for compiling the `node-bench` binary. We could rip out the `bin` folder from this repo and keep it in a different repo or in a fork of substrate but I think it will be easier to maintain if kept inside this repo.

**Review by Centril — APPROVED** [review]

**Pull request merged**